### PR TITLE
Use AreaStore

### PR DESCRIPTION
**Pull request by est31**
Use AreaStore data structure. See the minetest pr minetest/minetest#2898 which explains the speedups.

### CODE REVIEW

File: api.lua [hunk 1]
```diff
@@ -3,17 +3,48 @@
...
-		   pz >= ap1.z and pz <= ap2.z then
-			a[id] = area
+	if self.store then
+		local areas = self.store:get_areas_for_pos(pos, false, true)
+		for store_id, store_area in pairs(areas) do
+			local id = tonumber(store_area.data)
+			a[id] = self.areas[id]
+		end
+	else
+		for id, area in pairs(self.areas) do
```
Comment (kilbith) [inline api.lua:13]: Use a faster loop : 

```
for id = 1, #self.areas do
  local area = self.areas[id]
```

File: api.lua [hunk 2]
```diff
@@ -3,17 +3,48 @@
...
+	local res = {}
+	if self.store then
+		local areas = self.store:get_areas_in_area(pos1,
+			pos2, true, false, true)
+		for store_id, store_area in pairs(areas) do
+			local id = tonumber(store_area.data)
+			res[id] = self.areas[id]
+		end
+	else
+		for id, area in pairs(self.areas) do
```
Comment (kilbith) [inline api.lua]: Same here.

File: api.lua [hunk 3]
```diff
@@ -52,30 +83,23 @@ function areas:canInteractInArea(pos1, pos2, name, allow_open)
...
-		for id, area in pairs(self.areas) do
-			-- A little optimization: isAreaOwner isn't necessary
-			-- here since we're iterating through all areas.
-			if area.owner == name and
-					self:isSubarea(pos1, pos2, id) then
-				return true
-			end
+
+	local areas = self:getAreasForArea(pos1, pos2)
+	for id, area in pairs(areas) do
```
Comment (kilbith) [inline api.lua]: Here too.

File: internal.lua [hunk 4]
```diff
@@ -18,6 +18,23 @@ function areas:save()
...
 end
 
+local function populateStore(self)
+	if not AreaStore then
+		return
+	end
+	local store = AreaStore()
+	local store_ids = {}
+	store:reserve(#self.areas)
+	for id, area in pairs(areas.areas) do
```
Comment (kilbith) [inline internal.lua]: And here.

File: api.lua [hunk 5]
```diff
@@ -3,17 +3,48 @@
...
 	end
 	return a
 end
 
+function areas:getAreasForArea(pos1, pos2)
+	local res = {}
+	if self.store then
+		local areas = self.store:get_areas_in_area(pos1,
+			pos2, true, false, true)
+		for store_id, store_area in pairs(areas) do
```
Comment (kilbith) [inline api.lua]: Same here.

File: api.lua [hunk 6]
```diff
@@ -3,17 +3,48 @@
...
 	local px, py, pz = pos.x, pos.y, pos.z
-	for id, area in pairs(self.areas) do
-		local ap1, ap2 = area.pos1, area.pos2
-		if px >= ap1.x and px <= ap2.x and
-		   py >= ap1.y and py <= ap2.y and
-		   pz >= ap1.z and pz <= ap2.z then
-			a[id] = area
+	if self.store then
+		local areas = self.store:get_areas_for_pos(pos, false, true)
+		for store_id, store_area in pairs(areas) do
```
Comment (kilbith) [inline api.lua]: See below.

File: api.lua (hunk 1 above)
Comment (est31) [inline api.lua:13]: good idea

File: api.lua [hunk 7]
```diff
@@ -3,17 +3,48 @@
...
+			local id = tonumber(store_area.data)
+			res[id] = self.areas[id]
+		end
+	else
+		for id, area in pairs(self.areas) do
+			local p1, p2 = area.pos1, area.pos2
+			if (p1.x <= pos2.x and p2.x >= pos1.x) and
+					(p1.y <= pos2.y and p2.y >= pos1.y) and
+					(p1.z <= pos2.z and p2.z >= pos1.z) then
```
Comment (kilbith) [inline api.lua]: One tab please.

File: api.lua (hunk 1 above)
Comment (ShadowNinja) [inline api.lua:13]: This does seem to be faster, but only by about 1-8ms per million entries.  I don't think we need to optimize it that much.

Comment (kilbith) [inline api.lua]: According to what I see, it speeds up by 4x times on 50,000 entries.

File: api.lua [hunk 8]
```diff
@@ -3,17 +3,48 @@
...
 	local a = {}
 	local px, py, pz = pos.x, pos.y, pos.z
-	for id, area in pairs(self.areas) do
-		local ap1, ap2 = area.pos1, area.pos2
-		if px >= ap1.x and px <= ap2.x and
-		   py >= ap1.y and py <= ap2.y and
-		   pz >= ap1.z and pz <= ap2.z then
-			a[id] = area
+	if self.store then
+		local areas = self.store:get_areas_for_pos(pos, false, true)
```
Comment (ShadowNinja) [inline api.lua:7]: Didn't you change this to `_at_`?

**Comment by ShadowNinja** [comment]
It seems like this introduces _two_ area IDs: The one used in the AreaStore and the one used for everything else.  There should only be one.

File: internal.lua [hunk 9]
```diff
@@ -18,6 +18,23 @@ function areas:save()
 	file:close()
 end
 
+local function populateStore(self)
+	if not AreaStore then
```
Comment (ShadowNinja) [inline internal.lua]: Use `rawget(_G, "AreaStore")` to avoid a warning if it doesn't exist.

File: internal.lua [hunk 10]
```diff
@@ -18,6 +18,23 @@ function areas:save()
...
+		return
+	end
+	local store = AreaStore()
+	local store_ids = {}
+	store:reserve(#self.areas)
+	for id, area in pairs(areas.areas) do
+		local sid = store:insert_area(area.pos1,
+			area.pos2, dump(id))
```
Comment (ShadowNinja) [inline internal.lua]: `dump()` is for display, it's format is not designed to be read by anything other than a human and its output may not be consistent.  Use `tostring` here instead, it's also faster.

File: api.lua (hunk 8 above)
Comment (est31) [inline api.lua:7]: no

File: internal.lua (hunk 9 above)
Comment (est31) [inline internal.lua]: ok

File: internal.lua (hunk 10 above)
Comment (est31) [inline internal.lua]: ok

**Comment by est31** [comment]
@ShadowNinja the two area ids are done because areastore does its own id management (to allow easier use), but it may set the ids differently each time (the API doesn't guarantee that its the same, every time). I think this should be hidden completely from the users, and the ids should be constant over restarts.

**Comment by ShadowNinja** [comment]
@est31: Yes, but if that's the issue I think the areastore should handle the IDs better, or just provide a way to pass your own ID, bypassing its auto-chooser.

**Comment by kilbith** [comment]
Will it be hopefully merged next year ?

**Comment by davisonio** [comment]
:+1:

File: api.lua [hunk 11]
```diff
@@ -3,17 +3,48 @@
...
+			py >= ap1.y and py <= ap2.y and
+			pz >= ap1.z and pz <= ap2.z then
+				a[id] = area
+			end
 		end
 	end
 	return a
 end
 
+function areas:getAreasForArea(pos1, pos2)
```
Comment (ShadowNinja) [inline api.lua:25]: You have to make sure these positions are sorted.

File: api.lua (hunk 7 above)
Comment (ShadowNinja) [inline api.lua]: This actually checks if the passed area is entirely contained by the area you're iterating over, but we're interested in any areas that share even one node with the passed area.

**Comment by ShadowNinja** [comment]
I'll try to fix the issues I noticed and merge this now.

File: internal.lua [hunk 12]
```diff
@@ -18,6 +18,23 @@ function areas:save()
 	file:close()
 end
 
+local function populateStore(self)
+	if not rawget(_G, "AreaStore") then
+		return
+	end
+	local store = AreaStore()
+	local store_ids = {}
+	store:reserve(#self.areas)
```
Comment (ShadowNinja) [inline internal.lua:27]: The area list is actually a sparse array, so this won't work properly.

File: internal.lua [hunk 13]
```diff
@@ -29,6 +46,7 @@ function areas:load()
 	if type(self.areas) ~= "table" then
 		self.areas = {}
 	end
+	populateStore(self)
```
Comment (ShadowNinja) [inline internal.lua:49]: Should be done after the file's closed.

File: internal.lua [hunk 14]
```diff
@@ -46,6 +64,12 @@ function areas:add(owner, name, pos1, pos2, parent)
 	local id = findFirstUnusedIndex(self.areas)
 	self.areas[id] = {name=name, pos1=pos1, pos2=pos2, owner=owner,
 			parent=parent}
+	-- add to AreaStore
+	if self.store then
+		local sid = self.store:insert_area(pos1, pos2, dump(id))
```
Comment (ShadowNinja) [inline internal.lua:69]: This still uses `dump`.

File: api.lua [hunk 15]
```diff
@@ -52,30 +83,23 @@ function areas:canInteractInArea(pos1, pos2, name, allow_open)
...
-					(not name or not areas:isAreaOwner(id, name)) then
-				return false, id
-			end
+
+		-- Then check for intersecting (non-owned) areas.
+		-- Return if the area is closed or open areas aren't
+		-- allowed, and the area isn't owned.
+		if (not allow_open or not area.open) and
+				(not name or not self:isAreaOwner(id, name)) then
+			return false, id
```
Comment (ShadowNinja) [inline api.lua:102]: Combining these two loops into one like this won't work, as the function will return `false` if it finds a area that you don't own before finding a enclosing area that you do own.

**Comment by ShadowNinja** [comment]
8b0b8c1ad26afdae906e9fd6332218d79a33bd1d

**Pull request closed without merging**